core/internal/gethwrappers/abigen.go

```diff
@@ -54,4 +66,239 @@ func Abigen(a AbigenArgs) {
 		Exit("failure while building "+a.Pkg+" wrapper, stderr: "+
 			buildResponse.String(), err)
 	}
+
+	ImproveAbigenOutput(a.Out)
+}
+
+func ImproveAbigenOutput(path string) {
+	bs, err := ioutil.ReadFile(path)
+	if err != nil {
+		Exit("Error while improving abigen output", err)
+	}
+
+	fset, fileNode := parseFile(bs)
+	contractName := getContractName(fileNode)
+	fileNode = addAddressField(contractName, fileNode)
+	fileNode = replaceAnonymousStructs(contractName, fileNode)
+	bs = generateCode(fset, fileNode)
+	bs = writeAdditionalMethods(contractName, bs)
+	fset, fileNode = parseFile(bs)
+	fileNode = writeInterface(contractName, fileNode)
+	bs = generateCode(fset, fileNode)
+	bs = addHeader(bs)
+
+	err = ioutil.WriteFile(path, bs, 0600)
+	if err != nil {
+		Exit("Error while writing improved abigen source", err)
+	}
+}
+
+func parseFile(bs []byte) (*token.FileSet, *ast.File) {
+	fset := token.NewFileSet()
+	fileNode, err := parser.ParseFile(fset, "", string(bs), parser.AllErrors)
+	if err != nil {
+		Exit("Error while improving abigen output", err)
+	}
+	return fset, fileNode
+}
+
+func generateCode(fset *token.FileSet, fileNode *ast.File) []byte {
+	var buf bytes.Buffer
+	err := format.Node(&buf, fset, fileNode)
+	if err != nil {
+		Exit("Error while writing improved abigen source", err)
+	}
+	return buf.Bytes()
+}
+
+func getContractName(fileNode *ast.File) string {
+	// Grab the contract name. It's always the first type in the file.
+	var contractName string
+	astutil.Apply(fileNode, func(cursor *astutil.Cursor) bool {
+		x, is := cursor.Node().(*ast.TypeSpec)
+		if !is {
+			return true
+		}
+		if contractName == "" {
+			contractName = x.Name.Name
+		}
+		return false
+	}, nil)
+	return contractName
+}
+
+func addAddressField(contractName string, fileNode *ast.File) *ast.File {
```

Comment:
@connorwstein 
Going to pull in @spooktheducks here since he wrote this part. We had the goal of removing any intermediaries so that we could minimize our overhead when adding new contracts and using new contract functions / events. Agreed that we could be shooting ourselves in the foot if the upstream build changes too drastically. Imho though, the additions of the interface, named structs, and the Address() function made the consuming code and tests way cleaner, which may offset the risk of manipulating the output of abigen.

Reply:
yeah I can see the named structs being useful. The address one though seems simple to just put on the deviationChecker alongside the contract object and avoid that particular abigen output manipulation